eth/filters/filter_system.go

```diff
@@ -108,10 +121,36 @@ func NewEventSystem(mux *event.TypeMux, backend Backend, lightMode bool) *EventS
 		lightMode: lightMode,
 		install:   make(chan *subscription),
 		uninstall: make(chan *subscription),
+		txCh:      make(chan core.TxPreEvent, txChanSize),
+		logsCh:    make(chan []*types.Log, logsChanSize),
+		rmLogsCh:  make(chan core.RemovedLogsEvent, rmLogsChanSize),
+		chainCh:   make(chan core.ChainEvent, chainEvChanSize),
 	}
 
-	go m.eventLoop()
+	// Subscribe events
+	m.txSub = m.backend.SubscribeTxPreEvent(m.txCh)
+	if m.txSub == nil {
+		return nil
+	}
+	m.logsSub = m.backend.SubscribeLogsEvent(m.logsCh)
+	if m.logsSub == nil {
+		return nil
+	}
+	m.rmLogsSub = m.backend.SubscribeRemovedLogsEvent(m.rmLogsCh)
+	if m.rmLogsSub == nil {
+		return nil
+	}
+	m.chainSub = m.backend.SubscribeChainEvent(m.chainCh)
+	if m.chainSub == nil {
+		return nil
+	}
+	// TODO(rjl493456442): use feed to subscribe pending log event
+	m.pendingLogSub = m.mux.Subscribe(core.PendingLogsEvent{})
+	if m.pendingLogSub.Closed() {
+		return nil
+	}
```

Comment:
The `return nil`-s here are problematic, because they promote bad use. It says to the caller that they should check whether the return is `nil`, and handle it as an error. If we want to handle error cases, we need to return an `error`. If we don't want to handle error cases, we need to abort.

Returning `nil` is furthermore problematic because it doesn't clean up previous successful subscriptions.

In theory these code paths cannot get called, at least if the code initialization code is correct. As such we should either not care about them becoming `nil` (and panic if they do and find the culprit further up the stack), or if we want to handle things a bit more gracefully, we could do a `log.Crit` to abort.